### PR TITLE
feat(descriptors): handle validations

### PR DESCRIPTION
**Pull request by dcilke**
### Description

Ensures validations are handled when serializing a descriptor. This also handles converting manifest schema rules into actual Sanity rules.

### What to review

The validation serialization makes sense and there are not any edge case tests we are missing.

### Testing

Automated tests should cover most of it.

### Notes for release

Not required. This is for internal tooling.

### CODE REVIEW

**Comment by dcilke** [comment]
* **#10457** <a href="https://app.graphite.dev/github/pr/sanity-io/sanity/10457?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈 <a href="https://app.graphite.dev/github/pr/sanity-io/sanity/10457?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* `main`


<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.

File: packages/sanity/test/descriptors/schema.test.tsx [hunk 1]
```diff
@@ -11,7 +12,9 @@ import {
   type EncodedNamedType,
   type ObjectField,
 } from '../../../@sanity/schema/src/descriptors/types'
+import {ValidationError} from '../../../@sanity/schema/src/sanity/validation/ValidationError'
```
Comment (bjoerge) [inline packages/sanity/test/descriptors/schema.test.tsx]: any chance we could avoid imports outside of package boundaries here (and below)? Know its just tests, but it tends to create issues  down the road

File: packages/sanity/test/descriptors/schema.test.tsx [hunk 2]
```diff
@@ -400,7 +403,938 @@ describe('Base features', () => {
...
+            },
+          },
+        ],
+      })
+    })
+
+    // We expect this test to fail as extractWorkspaceManifest excludes messages which are not strings.
+    // See packages/sanity/src/_internal/manifest/extractWorkspaceManifest.tsx (transformValidation)
+    // for more details
+    test.fails('localized validation messages', () => {
```
Comment (bjoerge) [inline packages/sanity/test/descriptors/schema.test.tsx]: TIL about test.fails() – is this case something we expect to pass in at some point in the future?

Comment (dcilke) [reply packages/sanity/test/descriptors/schema.test.tsx]: Potentially. Ideally we would be convert the manifest types back into the source schema. We may go down that path, or we may focus our efforts/energy on moving towards a new paradigm.

The failing tests were a way of documenting the inconsistencies in behavior so we can keep track if/when they change.

Comment (bjoerge) [reply packages/sanity/test/descriptors/schema.test.tsx]: Nice nice

File: packages/sanity/test/descriptors/schema.test.tsx [hunk 3]
```diff
@@ -400,7 +403,938 @@ describe('Base features', () => {
...
+                  },
+                ],
+              },
+            ],
+          },
+        ],
+      })
+    })
+
+    // We expect this test to fail as extractWorkspaceManifest excludes validation rules
```
Comment (judofyr) [inline packages/sanity/test/descriptors/schema.test.tsx]: `convertType` does test more than just `extractWorkspaceManifest` though, so maybe we can split this up somehow? We do expect the result of `convertType` to actually match that object (and it does!). Maybe either move the edge-cases into `convertType` or have a separate `convertType`? 🤔

File: packages/@sanity/schema/src/descriptors/convert.ts [hunk 4]
```diff
@@ -289,11 +297,334 @@ function convertUnknown(
...
+  return typeDef.fields.some((field) => field.name === 'value')
+}
+
+function extractValueFromListOption(option: unknown, typeDef: Record<string, unknown>): unknown {
+  // If you define a `list` option with object items, where the item has a `value` field,
+  // we don't want to treat that as the value but rather the surrounding object
+  // This differs from the case where you have a title/value pair setup for a string/number, for instance
+  if (typeDef.jsonType === 'object' && hasValueField(typeDef)) return option
+
+  return (option as Record<string, unknown>).value === undefined
```
Comment (judofyr) [inline packages/@sanity/schema/src/descriptors/convert.ts]: Isn't this unsafe? E.g. if someone defines `options: {list: [null]}` then I think this will crash?

Comment (dcilke) [reply packages/@sanity/schema/src/descriptors/convert.ts]: Indeed! This was copied directly from packages/sanity/src/core/validation/util/normalizeValidationRules.ts. I'll add some more defenses.

File: packages/sanity/test/descriptors/schema.test.tsx (hunk 3 above)
Comment (dcilke) [reply packages/sanity/test/descriptors/schema.test.tsx]: I've taken a stab at this, let me know what you think. My other thought was to break the tests apart more having a suite for:
* schema -> descriptor
* schema -> manifest -> schema -> descriptor
* schema -> descriptor _vs_ schema -> manifest -> schema -> descriptor

File: packages/sanity/test/descriptors/schema.test.tsx (hunk 2 above)
Comment (dcilke) [reply packages/sanity/test/descriptors/schema.test.tsx]: Based on Holm's comment I've refactored the tests which made the failing calls more explicit (i.e. `expect(...).toThrow()`).

File: packages/@sanity/schema/src/descriptors/convert.ts [hunk 5]
```diff
@@ -289,11 +297,338 @@ function convertUnknown(
...
+  if (isObject(val) && '$$typeof' in val && 'type' in val && 'props' in val) {
     const {type, props} = val
     const strType = typeof type === 'function' ? type.name : type
     if (typeof strType !== 'string') return undefined
     return {__type: 'jsx', type: strType, props: convertUnknown(props) as EncodableObject}
   }
   return undefined
 }
+
+// maybeValidations attempts to serialize the validations of a type. Note: we need the who type object and not
```
Comment (jordanl17) [inline packages/@sanity/schema/src/descriptors/convert.ts]:
```suggestion
// maybeValidations attempts to serialize the validations of a type. Note: we need the whole type object and not
```

File: packages/@sanity/schema/src/descriptors/convert.ts [hunk 6]
```diff
@@ -289,11 +297,338 @@ function convertUnknown(
...
+      impliedRules.push({
+        type: 'email',
+      })
+      break
+    default:
+    // Do nothing
+  }
+
+  // Shortcut
+  if (!('validation' in obj) || !obj.validation) {
```
Comment (jordanl17) [inline packages/@sanity/schema/src/descriptors/convert.ts:360]: Q: Are both conditions here necessary. If just wanting this branch to capture when `obj.validation` is falsy then can probably just go with `!obj.validation`?

File: packages/@sanity/schema/src/sanity/createSchemaFromManifestTypes.ts [hunk 7]
```diff
@@ -1,25 +1,292 @@
...
+  // Convert ManifestValidationGroup[] to Rule functions
+  const manifestValidation = Array.isArray(val.validation) ? val.validation : [val.validation]
+
+  val.validation = manifestValidation
+    .map((group) => {
+      if (!isObject(group)) return undefined
+
+      return (baseRule: IRule) => {
+        let rule = baseRule
+        const level = 'level' in group ? group.level : undefined
```
Comment (jordanl17) [inline packages/@sanity/schema/src/sanity/createSchemaFromManifestTypes.ts:79]: Nit: Simpler to just do:
```suggestion
        const level = group.level
```
?

**Review by jordanl17 — APPROVED** [review]
Syntactically looks good. The nits are non-blocking, but could help simplify some of the value getting

File: packages/@sanity/schema/src/descriptors/convert.ts (hunk 6 above)
Comment (dcilke) [reply packages/@sanity/schema/src/descriptors/convert.ts:360]: This is a Typescriptism. Essentially:
* `obj` -> `{}`
* ``validation` in obj` -> `{validation: unknown}`
* `!obj.validation` -> `{validation: <not falsy>}`

File: packages/@sanity/schema/src/sanity/createSchemaFromManifestTypes.ts (hunk 7 above)
Comment (dcilke) [reply packages/@sanity/schema/src/sanity/createSchemaFromManifestTypes.ts:79]: True, but since we are using `unknown` Typescript requires stricter property fetching.
